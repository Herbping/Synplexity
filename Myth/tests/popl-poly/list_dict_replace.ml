type nat = 
  | Z
  | S of nat

type 'a list =
  | Nil
  | Cons of 'a * 'a list

type bool =
  | False
  | True

type 'a option =
  | Some of 'a
  | None

let rec list_dict_replace : ('a -> 'a -> bool) -> 'a -> 'b -> ('a * 'b) list -> ('a * 'b) list |>
  /\(a1 -> a1 -> True,
     a1 -> a2 -> False,
     a2 -> a1 -> False) ->
  a1 -> b2 ->
  /\(
        [] -> [],
        [(a1, 'b)] -> [(a1, b2)],
        [(a2, b1)] -> [(a2, b1)],
        [(a2, b1); (a1, 'b)] -> [(a2, b1); (a1, b2)],
        [(a1, 'b); (a2, b1)] -> [(a1, b2); (a2, b1)]
    )
= ?
