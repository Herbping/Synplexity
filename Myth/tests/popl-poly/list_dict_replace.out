(* Generated Wed Nov 11 22:02:32 2015 *)
fix f1 (f2:'a -> 'a -> bool) : 'a -> 'b -> 'a * 'b list -> 'a * 'b list ->
  fix f3 (p1:'a) : 'b -> 'a * 'b list -> 'a * 'b list ->
    fix f4 (p2:'b) : 'a * 'b list -> 'a * 'b list ->
      fix f5 (l1:'a * 'b list) : 'a * 'b list ->
        match l1 with
        | Nil u1 -> Nil
        | Cons t1 -> match (f2 p1) (#1 (#1 t1)) with
          | False u1 -> Cons ((#1 (#1 t1), #2 (#1 t1)), f5 (#2 t1))
          | True u1 -> Cons ((p1, p2), #2 t1)
